src/opaque_closure.c

```diff
@@ -57,44 +41,70 @@ static jl_opaque_closure_t *new_opaque_closure(jl_tupletype_t *argt, jl_value_t
-    JL_GC_POP();
-    oc->source = source;
-    oc->captures = captures;
-    oc->specptr = NULL;
-    if (!ci) {
-        oc->invoke = (jl_fptr_args_t)jl_interpret_opaque_closure;
-    } else {
-        jl_callptr_t invoke = jl_atomic_load_acquire(&ci->invoke);
-        if (invoke == jl_fptr_interpret_call) {
-            oc->invoke = (jl_fptr_args_t)jl_interpret_opaque_closure;
+    jl_fptr_args_t invoke = (jl_fptr_args_t)jl_interpret_opaque_closure;
+    void *specptr = NULL;
+
+    if (ci) {
+        invoke = (jl_fptr_args_t)jl_atomic_load_relaxed(&ci->invoke);
+        specptr = jl_atomic_load_relaxed(&ci->specptr.fptr);
+        if (invoke == (jl_fptr_args_t) jl_fptr_interpret_call) {
+            invoke =(jl_fptr_args_t) jl_interpret_opaque_closure;
         }
-        else if (invoke == jl_fptr_args) {
-            oc->invoke = jl_atomic_load_relaxed(&ci->specptr.fptr1);
+        else if (invoke == (jl_fptr_args_t)jl_fptr_args) {
+            invoke = (jl_fptr_args_t)specptr;
         }
-        else if (invoke == jl_fptr_const_return) {
-            oc->invoke = (jl_fptr_args_t)jl_fptr_const_opaque_closure;
-            oc->captures = ci->rettype_const;
+        else if (invoke == (jl_fptr_args_t)jl_fptr_const_return) {
+            invoke = (jl_fptr_args_t)jl_fptr_const_opaque_closure;
+            captures = ci->rettype_const;
         }
-        else {
-            oc->invoke = (jl_fptr_args_t) invoke;
+
+        if (!jl_subtype(rt_lb, ci->rettype)) {
+            // If we're not allowed to generate a specsig with this, rt, fall
+            // back to the invoke wrapper. We could instead generate a specsig->specsig
+            // wrapper, but lets leave that for later.
+            specptr = NULL;
+            selected_rt = jl_type_intersection(rt_lb, ci->rettype);
+        } else {
+            selected_rt = ci->rettype;
         }
     }
+
+    jl_value_t *oc_type JL_ALWAYS_LEAFTYPE = jl_apply_type2((jl_value_t*)jl_opaque_closure_type, (jl_value_t*)argt, selected_rt);
+    JL_GC_PROMISE_ROOTED(oc_type);
+
+    if (!specptr) {
+        // Generate a specsig wrapper that just calls ->invoke
+        jl_method_t *generic_method = (jl_method_t*)jl_methtable_lookup(jl_opaque_closure_typename->mt, (jl_value_t*)jl_anytuple_type, world);
+
+        sigtype = jl_argtype_with_function_type((jl_value_t*)oc_type, (jl_value_t*)argt);
+        jl_method_instance_t *mi_generic = jl_specializations_get_linfo(generic_method, sigtype, jl_emptysvec);
+
+        // OC wrapper methods are not world dependent
+        ci = jl_get_method_inferred(mi_generic, selected_rt, 1, ~(size_t)0);
+        if (!jl_atomic_load_acquire(&ci->invoke))
+            jl_generate_fptr_for_oc_wrapper(ci);
+        specptr = jl_atomic_load_relaxed(&ci->specptr.fptr);
+    }
+    jl_opaque_closure_t *oc = (jl_opaque_closure_t*)jl_gc_alloc(ct->ptls, sizeof(jl_opaque_closure_t), oc_type);
+    oc->source = source;
```

Comment:
Is it reasonable to use inferred source of `ci` here? For top-level opaque closures they are inferred and their return type are computed (since we specify `do_compile=1`) but their source are uninferred ones.
```julia
julia> opc = Base.Experimental.@opaque x::Float64 -> x + 1.0
(::Float64)::Float64->◌

julia> @ccall jl_uncompress_ir(opc.source::Any, C_NULL::Ptr{Cvoid}, opc.source.source::Any)::Any
CodeInfo(
    @ REPL[1]:1 within `opaque closure`
1 ─ %1 = x + 1.0
└──      return %1
)
```

Reply:
The `source` here is a method of which you're looking at the uninferred source. We may want to update the reflection functions to see if we can make them find the correct specialization, but in general it's not guaranteed that opaque closures are introspectable.